contracts/eden/src/eden-micro-chain.cpp

```diff
@@ -0,0 +1,588 @@
+                                                     uint32_t token_account_low,
+                                                     uint32_t token_account_high,
+                                                     uint32_t atomic_account_low,
+                                                     uint32_t atomic_account_high,
+                                                     uint32_t atomicmarket_account_low,
+                                                     uint32_t atomicmarket_account_high)
+{
+   __wasm_call_ctors();
+   eden_account.value = (uint64_t(eden_account_high) << 32) | eden_account_low;
+   token_account.value = (uint64_t(token_account_high) << 32) | token_account_low;
+   atomic_account.value = (uint64_t(atomic_account_high) << 32) | atomic_account_low;
+   atomicmarket_account.value =
+       (uint64_t(atomicmarket_account_high) << 32) | atomicmarket_account_low;
+}
+
+[[clang::export_name("allocateMemory")]] void* allocateMemory(uint32_t size)
+{
+   return malloc(size);
+}
+[[clang::export_name("freeMemory")]] void freeMemory(void* p)
+{
+   free(p);
+}
+
+std::variant<std::string, std::vector<char>> result;
+[[clang::export_name("getResultSize")]] uint32_t getResultSize()
+{
+   return std::visit([](auto& data) { return data.size(); }, result);
+}
+[[clang::export_name("getResult")]] const char* getResult()
+{
+   return std::visit([](auto& data) { return data.data(); }, result);
+}
+
+template <typename T>
+void dump(const T& ind)
+{
+   printf("%s\n", eosio::format_json(ind).c_str());
+}
+
+namespace boost
+{
+   BOOST_NORETURN void throw_exception(std::exception const& e)
+   {
+      eosio::detail::assert_or_throw(e.what());
+   }
+   BOOST_NORETURN void throw_exception(std::exception const& e, boost::source_location const& loc)
+   {
+      eosio::detail::assert_or_throw(e.what());
+   }
+}  // namespace boost
+
+struct by_id;
+struct by_pk;
+struct by_invitee;
+
+template <typename T, typename... Indexes>
+using mic = boost::
+    multi_index_container<T, boost::multi_index::indexed_by<Indexes...>, chainbase::allocator<T>>;
+
+template <typename T>
+using ordered_by_id = boost::multi_index::ordered_unique<  //
+    boost::multi_index::tag<by_id>,
+    boost::multi_index::member<T, typename T::id_type, &T::id>>;
+
+template <typename T>
+using ordered_by_pk = boost::multi_index::ordered_unique<  //
+    boost::multi_index::tag<by_pk>,
+    boost::multi_index::const_mem_fun<
+        T,
+        eosio::remove_cvref_t<typename eosio::member_fn<decltype(&T::pk)>::return_type>,
+        &T::pk>>;
+
+template <typename T>
+using ordered_by_invitee = boost::multi_index::ordered_unique<  //
+    boost::multi_index::tag<by_invitee>,
+    boost::multi_index::const_mem_fun<
+        T,
+        eosio::remove_cvref_t<typename eosio::member_fn<decltype(&T::invitee)>::return_type>,
+        &T::invitee>>;
+
+auto available_pk(const auto& table, const auto& first)
+    -> eosio::remove_cvref_t<decltype(table.begin()->pk())>
```

Comment:
I don't want to unnecessarily restrict `available_pk` to tables with `uint64_t` pks.

Reply:
If you just needed an auto-incrementing counter, then the chainbase object ID already provides it.  The reason that `available_primary_key` is needed is to make sure that the induction ids exactly match the ids used by the contract.  Making it generic serves no purpose, since contracts only have `uint64_t` primary keys.

Reply:
changed